src/tool/subcommands/api_cmd.rs

```diff
@@ -746,6 +746,31 @@ fn websocket_tests() -> Vec<RpcTest> {
     vec![test]
 }
 
+// Sanity check `ApiInfo` and `ApiTestFlags` args
+fn validate_protocols(
+    forest: &ApiInfo,
+    lotus: &ApiInfo,
+    config: &ApiTestFlags,
+) -> anyhow::Result<()> {
+    let a = forest.multiaddr.clone().pop().map(|p| p.tag());
+    let b = lotus.multiaddr.clone().pop().map(|p| p.tag());
+
+    let is_valid = match (a, b) {
+        (Some("http"), Some(x)) if x != "http" => false,
+        (Some("ws"), Some(x)) if x != "ws" => false,
+        (Some(x), Some("http")) if x != "http" => false,
+        (Some(x), Some("ws")) if x != "ws" => false,
+        (Some("http"), Some("http")) if config.use_websocket => {
+            bail!("can't use http communication when `--ws` flag is used")
```

Comment:
In general, is `--ws` flag needed? It seems it could be derived from the `ApiInfo` (i.e., set to `true` if both endpoints use websockets). Am I missing something?

Reply:
If we derived the `--ws` flag from the `ApiInfo`, that means we will have to specify each time the `ApiInfo` for both Forest (e.g., `--forest /ip4/127.0.0.1/tcp/2345/ws`) *and* Lotus if we want to use websocket communication. `ApiInfo` is optional and has a default value.

If you think it could be more evident to the user, then I'm okay with removing the flag.

Reply:
I'd probably go with an explicit approach without redundancy, but if you want to keep it as it is, I'm okay with that.